### ISSUE TITLE
SPIKE: Kafka Streams App Learning

### ISSUE THREAD

**Issue by EvanParish** [department-of-veterans-affairs/notification-api#1279]
### User Story(ies)
**As a** VA Notify Engineer
**I want** to know more about how a Kafka streams app works  
**So that** we can build our integration when we are ready.

### Additional Info and Resources
Check Jacob's notes on a simple [Kafka streaming app](https://github.com/department-of-veterans-affairs/vanotify-team/blob/master/Architecture/kafka/kafka_libraries.md).

There are various python packages available for reading events from a Kafka stream. We should look into each and decide on the best approach. The following are the ones currently being maintained.
- [confluent-kafka](https://github.com/confluentinc/confluent-kafka-python) - a package maintained by the developers of Kafka
- ~[faust-streaming](https://github.com/faust-streaming/faust) - a more up-to-date version of faust~

## Engineering Checklist
_**Required**_

- [x] Write a test app to read data from a Kafka stream
  - [x] read data from an offset or timestamp
  - [x] deserialize data (AVRO)
  - [x] handle possible errors
  - [x] connect to a Kafka topic over SSL
- [x] Document understanding of how to connect to BIP's Kafka topic and read events from it

**Note**: This test app should be written outside of the repo somewhere. If you check Jacob's notes above, you can see some example code to get you started. He was running this code in the dev-test-instance EC2.

## Acceptance Criteria
_**Required**_
- A test app has been created. 
- We understand the main areas and concerns associated with connecting to a Kafka topic
  - How to connect via ssl (usually you just need to connect with the url of the kafka instance but when dealing with ssl, you need to provide the cert data)
  - How to start at different offsets (date or offset)
  - How to deserialize a stream (AVRO)
  - How to handle unexpected end of stream (sometimes you can run into an error like with serialization that shuts down the consumer, it's good to know how to handle that gracefully and/or recover)
- Successful connection to BIP Kafka topic 
  1. able to read events from the topic (byte array)
  2. able to deserialize events from the topic

## QA Considerations
_For QA to populate. Leave blank if QA is not applicable on this ticket._

## Potential Dependencies
_Leave blank if n/a_

## Out of Scope
_Leave blank if n/a_

**Comment by mjones-oddball** [comment]
Hey team! Please [add your planning poker estimate](https://app.zenhub.com/workspaces/VA-Notify-620d21369d810a00146ed9c8/issues/department-of-veterans-affairs/notification-api/1279?planning-poker) with Zenhub @cris-oddball @EvanParish @justaskdavidb2 @k-macmillan @kalbfled @ldraney @nikolai-efimov

**Comment by mjones-oddball** [comment]
Open question: Where would this POC code go?
- We are going to update the doc Jacob started: https://github.com/department-of-veterans-affairs/vanotify-team/blob/master/Architecture/kafka/kafka_libraries.md
- Actual code could be stored in Dev test instance with coordination from team

**Comment by nikolai-efimov** [comment]
Does anyone know if we'll be using a schema registry to share AVRO schemas?

**Comment by nikolai-efimov** [comment]
@cris-oddball , we need the schema to deserialize the message. It can either be shared via a registry or it comes in as a header in a message. There are nuances in handling messages depending on which approach is used.

**Comment by cris-oddball** [comment]
@nikolai-efimov I hear you. They provided that schema to us, so I believe we can use it to de-serialize. I will try to get an example event off the topic for you to see if it comes in as a header.

**Comment by nikolai-efimov** [comment]
@EvanParish @mjones-oddball @tabinda-syed  I'm pretty much done with the app and need help with two things to wrap this up:

 1. Info BIP's Kafka topic (certs, keystores, schema sharing method (hard coded, registry, message header), server address, etc)
 2. Test it in company's environment

**Comment by nikolai-efimov** [comment]
@cris-oddball , I've tested things with the schema you provided (thank you for that!), it all works well. Serialization and deserialization are tested. As is it makes it a hardcoded schema for now, however, Kafka provides ways of receiving updated versions of schema either via a registry or in the message.

**Comment by nikolai-efimov** [comment]
@EvanParish , @cris-oddball , I got the following reply from Thomas in slack. 
"Hey Nikolai. It's not just you guys. Every External Tenant I've been working with is seeing the same Error. I have already requested help in resolving the issue. We're cautioning against updating the Group ID setting until we've had a chance to look into this more, but in the mean time for verification the group that we have listed for VA Notify is "vanotify-dev".
